### PR TITLE
DOC: Clarify when RandomState re-seeding is done.

### PR DESCRIPTION
**Pull request by ozabluda**

### CODE REVIEW

File: numpy/random/mtrand/mtrand.pyx [hunk 1]
```diff
@@ -619,9 +619,10 @@ cdef class RandomState:
         Random seed used to initialize the pseudo-random number generator.  Can
         be any integer between 0 and 2**32 - 1 inclusive, an array (or other
         sequence) of such integers, or ``None`` (the default).  If `seed` is
-        ``None``, then `RandomState` will try to read data from
-        ``/dev/urandom`` (or the Windows analogue) if available or seed from
-        the clock otherwise.
+        ``None``, then, every time 'RandomState' is initialized or
```
Comment (charris) [inline numpy/random/mtrand/mtrand.pyx]: What other times is ``` `RandomState` ``` initialized. Note single backticks. Also, there should not be a comma after "then".

**Comment by charris** [comment]
If you add `[ci skip]` to the commit message body, the tests won't be run when you are only changing documentation.

File: numpy/random/mtrand/mtrand.pyx (hunk 1 above)
Comment (ozabluda) [reply numpy/random/mtrand/mtrand.pyx]: Fixed.

File: numpy/random/mtrand/mtrand.pyx [hunk 2]
```diff
@@ -619,9 +619,10 @@ cdef class RandomState:
         Random seed used to initialize the pseudo-random number generator.  Can
         be any integer between 0 and 2**32 - 1 inclusive, an array (or other
         sequence) of such integers, or ``None`` (the default).  If `seed` is
-        ``None``, then `RandomState` will try to read data from
-        ``/dev/urandom`` (or the Windows analogue) if available or seed from
-        the clock otherwise.
+        ``None``, then every time `RandomState` is initialized or
+        ``numpy.random.seed(None)`` is called, `RandomState` will try to read
```
Comment (rkern) [inline numpy/random/mtrand/mtrand.pyx]: This is written oddly for this part of the docstring, which is about the constructor for `RandomState`. `numpy.random.seed()` isn't relevant, and "every time `RandomState` is initialized" is implicit in the context, because it is *only* talking about times that `RandomState` is initialized. If you want to document the behavior of `numpy.random.seed()`, do it in the `seed()` method.

I think the original wording is more correct and understandable.

Comment (ozabluda) [reply numpy/random/mtrand/mtrand.pyx]: The reason I did it in `RandomState`, is because `numpy.random.seed()` is currently documented 'by reference' to `RandomState` (maybe to avoid complede duplication of the docs for the seed parameter?):
https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.random.seed.html

Initially, my goal was only to document the re-seeding behavior of `numpy.random.seed(None)`, and I added "every time RandomState is initialized..." only to preserve correctness. The wording is indeed awkward - the best I could do.

Comment (rkern) [reply numpy/random/mtrand/mtrand.pyx]: I'm not sure I follow what's confusing (I don't doubt that it *is* confusing to someone; I'm just trying to understand exactly how the current text failed so we can get to the best fix).

This is a description of how an argument (either to the constructor or the `seed()` method) is interpreted. No matter which way you come to it, the context of that lookup tells you when it's relevant. If you start by reading the `RandomState` docstring, the `Parameters` describes the arguments to the constructor. If you start by reading the `RandomState.seed()` docstring, you follow the referral to see how it interprets the `seed` argument when it is called. The only reason you are reading the text in the first place is to know what `RandomState()` or `RandomState.seed()` is doing; you don't need to recapitulate that `RandomState()` or `RandomState.seed()` are the calls that interpret the `seed` argument.

Comment (ozabluda) [reply numpy/random/mtrand/mtrand.pyx]: The initial confusion (to me) I was trying to resolve (for myself) was:

If I call `numpy.random.seed(None)`, does it re-seed or not. In particular, is it idempotent (i.e. is calling it twice has the same semantics as calling it once). I resolved it for myself by seeing extra access to /dev/urandom in strace.

You can see an example of my confusion in 
https://github.com/fchollet/keras/pull/7239#pullrequestreview-65999280

Maybe it would be clearer if the bulk of the docs for the seed argument is moved from `RandomState` to `numpy.random.seed()`, with `RandomState` docs saying that the initializer actually calls `numpy.random.seed()` (which it currently doesn't say).

Comment (rkern) [reply numpy/random/mtrand/mtrand.pyx]: If the contents of the `seed` argument documentation moved verbatim to the `RandomState.seed()` method docstring, would you still have been confused? The reference back to the `RandomState` docstring is intended mean nothing more or less than that. Was there something in either text that prompted the idea that `RandomState.seed(None)` would be a no-op rather than initializing from `/dev/urandom`?

I don't approve of moving the documentation to `RandomState.seed()`. No one should ever call `numpy.random.seed()` or `RandomState.seed()` (except very few weird cases to work around broken code; the `keras` PR you linked isn't one of those, actually). Everyone should be instantiating `RandomState()` instances and passing them around and never reseeding them. `scikit-learn` is a good example of doing this consistently.

Comment (ozabluda) [reply numpy/random/mtrand/mtrand.pyx]: Before I reply, just want to make sure I understood you correctly. The Keras code is currently wrong. Because 
>it calls numpy.random.seed() instead of instantiating RandomState() instances and passing them around and never reseeding them

Comment (rkern) [reply numpy/random/mtrand/mtrand.pyx]: I'd probably say "sub-optimally designed" than "wrong" per se, but generally yes. It is probably *possible* to use the Keras APIs correctly and get reproducible results, but it's quite easy to misuse them. For example, there are a lot of places where you can set a seed through Keras APIs, through several different preprocessing options, dataset-loading, and layer initializations. All of them reseed the global `numpy.random` `RandomState`. If you were to pass a seed to more than one of these Keras APIs, you will re-initialize the `RandomState` back to an old state and reproduce the same bits that you had earlier. This can introduce bad auto-correlations. As a general rule for any PRNG, you always want to initialize exactly once and use that generator without reinitializing.

It's possible that there is a good reason for using the global `numpy.random` `RandomState`, possibly having to do with the interaction with the backends which might need to coordinate with a GPU PRNG or whatever. But in that case, the only correct way to use the `numpy.random` functions is to tell users that they need to call `numpy.random.seed()` at the beginning of their script exactly once and remove all ways to set a seed from the Keras API.

I'm not going to go through every use of `numpy.random.seed()` in `keras`, but I can say for sure that the dataset-loading code is definitely "wrong" and can be easily fixed without making any problem for the backends (if indeed there is an issue with coordinating with the backends). For example, we can rewrite the following lines like so: https://github.com/fchollet/keras/blob/master/keras/datasets/boston_housing.py#L27-L30

```
prng = np.random.RandomState(seed)
prng.shuffle(x)
# Repeat the same shuffling of y as x.
prng = np.random.RandomState(seed)
prng.shuffle(y)
```

Comment (charris) [reply numpy/random/mtrand/mtrand.pyx]: Can this be closed?

Comment (charris) [reply numpy/random/mtrand/mtrand.pyx]: @rkern Note that `np.random.seed` is used quite a lot in tests, especially in the testing of random. I assume that your comments are more intended for use in actual simulations and such?

@ozabluda I'm going to close this for now, please reopen if you see a continuing need.

Comment (rkern) [reply numpy/random/mtrand/mtrand.pyx]: No, I meant all code. `np.random.seed()` should only be used when working around broken code that only uses the `np.random` convenience aliases. And if you can control that code, it should be fixed to use a `RandomState` instance instead.

The tests in `numpy.random.tests` are not constructed according to best practice. They really should have been written to test the methods of `RandomState`, not the convenience aliases.
